### PR TITLE
Pluggable latex

### PR DESCRIPTION
**Pull request by ian-r-rose**
Adresses #2280. This adds a `latexTypesetter` property to the application rendermime. It defaults to an implementation using MathJax, but can be swapped for one using another renderer (such as KaTeX). A proof-of-concept extension for KaTeX is [here](https://github.com/ian-r-rose/jupyterlab-katex).

cc @jasongrout @mpacer

### CODE REVIEW

File: packages/rendermime/src/renderers.ts [hunk 1]
```diff
@@ -135,6 +138,11 @@ namespace renderHTML {
      * Whether the node should be typeset.
      */
     shouldTypeset: boolean;
+
+    /**
+     * The Latex typesetter for the application.
```
Comment (blink1073) [inline packages/rendermime/src/renderers.ts]: Perhaps, "The LaTeX typesetter" here as well?

Comment (blink1073) [reply packages/rendermime/src/renderers.ts]: (same in other places in this file)

File: packages/rendermime/src/latex.ts [hunk 2]
```diff
@@ -9,20 +9,86 @@
...
 /**
+ * The MathJax Typesetter.
+ */
+export
+class MathJaxTypesetter implements IRenderMime.ILatexTypesetter {
+  /**
+   * Construct the typesetter.
+   */
+  constructor() {
+    this._init();
```
Comment (blink1073) [inline packages/rendermime/src/latex.ts]: Perhaps not init here so we don't call this unless we actually render something.

File: packages/rendermime/src/renderers.ts (hunk 1 above)
Comment (ian-r-rose) [reply packages/rendermime/src/renderers.ts]: I agree. Probably not for type and variable names, though?

Comment (blink1073) [reply packages/rendermime/src/renderers.ts]: agreed

File: packages/codemirror/src/codemirror-ipythongfm.ts [hunk 3]
```diff
@@ -13,7 +13,7 @@ import 'codemirror/addon/mode/multiplex';
  * Define an IPython GFM (GitHub Flavored Markdown) mode.
  *
  * Is just a slightly altered GFM Mode with support for latex.
```
Comment (mpacer) [inline packages/codemirror/src/codemirror-ipythongfm.ts]: Could you also fix the capitalisation there too?

**Comment by mpacer** [comment]
I'm a little confused as to how this allows both the async and synchronous processing models that MathJax and KaTeX embody (respectively). Could the docs explain that in case someone wants to extend this for a third math processing engine?

**Comment by ian-r-rose** [comment]
Basically, the `typeset` API does not care whether the typesetting is done asynchronously (as with MathJax), or synchronously (as with KaTeX). Because it does not return a Promise, however, there is no way to hook into the point at which MathJax is done rendering.

**Comment by blink1073** [comment]
Great, thanks!

**Comment by scottdraves** [comment]
this is awesome

**Pull request merged**